File: hackerrank/challenges/RoadAndLibraries.js
// Complete the roadsAndLibraries function below.
function roadsAndLibraries(n, c_lib, c_road, cities) {
    let graph = new Array(n + 1)
    for (let i = 1; i <= n; i++) {
        graph[i] = []
    }
    let visited = new Array(n + 1)
    let queue = []

    cities.forEach(function ([from, to]) {
        graph[from].push(to)
        graph[to].push(from)
    })

    let result = 0;
    graph.forEach(function (connections, city) {
        if (visited[city]) return;

        result += c_lib;
        visited[city] = true

        connections.forEach(function (to) {
            queue.push(to)
        })

        while (queue.length !== 0) {
            let node = queue.shift()

            if (visited[node]) continue;

            visited[node] = true;
            result += (c_lib > c_road) ? c_road : c_lib

            graph[node].forEach(function (nodeTo) {
                queue.push(nodeTo)
            })
        }
    })

    return result;
}

let cities = [
    [1, 2],
    [2, 3],
    [3, 1]
]
console.log(roadsAndLibraries(3, 2, 1, cities)) // 4


cities = [
    [1, 2],
    [2, 3],
    [2, 4],
    [1, 3],
    [5, 6]
]
console.log(roadsAndLibraries(6, 2, 5, cities)) // 12

File: hackerrank/challenges/RoadAndLibraries.java
package challenges;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class RoadAndLibraries {
    static long roadsAndLibraries(int n, int c_lib, int c_road, int[][] cities) {
        List<Integer>[] graph = new LinkedList[n + 1];
        for (int i = 1; i <= n; i++) {
            graph[i] = new LinkedList<>();
        }
        boolean[] visited = new boolean[n + 1];
        Queue<Integer> queue = new LinkedList<>();

        for (int[] connection : cities) {
            graph[connection[0]].add(connection[1]);
            graph[connection[1]].add(connection[0]);
        }

        long result = 0;
        for (int i = 1; i <= n; i++) {
            if (visited[i]) continue;

            result += c_lib;
            visited[i] = true;

            queue.addAll(graph[i]);

            while (!queue.isEmpty()) {
                int node = queue.poll();
                if (visited[node]) continue;

                visited[node] = true;
                result += Math.min(c_lib, c_road);

                queue.addAll(graph[node]);
            }
        }

        return result;
    }
}
